src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/AutoDisenchanter.java

```diff
@@ -0,0 +1,140 @@
+package io.github.thebusybiscuit.slimefun4.implementation.items.electric.machines.enchanting;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.EnchantmentStorageMeta;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.bukkit.inventory.meta.Repairable;
+
+import io.github.thebusybiscuit.cscorelib2.inventory.InvUtils;
+import io.github.thebusybiscuit.slimefun4.api.events.AutoDisenchantEvent;
+import me.mrCookieSlime.Slimefun.Lists.RecipeType;
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.AContainer;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.MachineRecipe;
+import me.mrCookieSlime.Slimefun.api.SlimefunItemStack;
+import me.mrCookieSlime.Slimefun.api.inventory.BlockMenu;
+
+/**
+ * The {@link AutoDisenchanter}, in contrast to the {@link AutoEnchanter}, removes
+ * {@link Enchantment Enchantments} from a given {@link ItemStack} and transfers them
+ * to a book.
+ * 
+ * @author TheBusyBiscuit
+ * @author Walshy
+ * @author poma123
+ * 
+ * @see AutoEnchanter
+ *
+ */
+public class AutoDisenchanter extends AContainer {
+
+    @ParametersAreNonnullByDefault
+    public AutoDisenchanter(Category category, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
+        super(category, item, recipeType, recipe);
+    }
+
+    @Override
+    public ItemStack getProgressBar() {
+        return new ItemStack(Material.DIAMOND_CHESTPLATE);
+    }
+
+    @Override
+    protected MachineRecipe findNextRecipe(BlockMenu menu) {
+        Map<Enchantment, Integer> enchantments = new HashMap<>();
+
+        for (int slot : getInputSlots()) {
+            ItemStack item = menu.getItemInSlot(slot);
+
+            if (!isDisenchantable(item)) {
+                return null;
+            }
+
+            AutoDisenchantEvent event = new AutoDisenchantEvent(item);
+            Bukkit.getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return null;
+            }
+
+            ItemStack target = menu.getItemInSlot(slot == getInputSlots()[0] ? getInputSlots()[1] : getInputSlots()[0]);
+
+            // Disenchanting
+            if (target != null && target.getType() == Material.BOOK) {
+                int amount = 0;
+
+                for (Map.Entry<Enchantment, Integer> entry : item.getEnchantments().entrySet()) {
```

Comment:
I don't think a for-loop is necesary here, why not do this:

```java
Map<Enchantment, Integer> itemEnchantments = item.getEnchantments(); 
int amount = itemEnchantments.size();
enchantments.putAll(itemEnchantments);
```

Reply:
That file isn't part of this pull request though, it was just simply relocated.

Reply:
Ah, I see. I guess that will be for another PR then.